### PR TITLE
PROZHP-142 - Ersatz für Google Analytics auf ProZ

### PR DESCRIPTION
**Pull request by fpichler**
https://synjira.atlassian.net/browse/PROZHP-142

### CODE REVIEW

File: src/main/webapp/resources/celDynJS/analytics/analyticsQueue.mjs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  #analyticsEventQueue;
+  #analyticsPriorizedEventQueue;
+
+  constructor() {
+    document.addEventListener('visibilitychange', () => {
+      if (document.visibilityState === "hidden") {
+        this._sendMissedHits();
+      }
+    });
+	  this.#analyticsEventQueue = [];
```
Comment (ebeutler) [inline src/main/webapp/resources/celDynJS/analytics/analyticsQueue.mjs]:
```suggestion
    this.#analyticsEventQueue = [];
```

File: src/main/webapp/resources/celRes/analyticsQueue.js [hunk 2]
```diff
@@ -21,61 +21,44 @@
...
+	  _me._addBind = _me.add.bind(_me);
+	  _me._priorityAddBind = _me.priorityAdd.bind(_me);
+	  _me._sizeBind = _me.size.bind(_me);
+	  _me._lowPrioritySizeBind = _me.lowPrioritySize.bind(_me);
+	  _me._highPrioritySizeBind = _me.highPrioritySize.bind(_me);
+	  _me._sendHitsBind = _me._sendHits.bind(_me);
+	  _me._getPrioMsgPartBind = _me._getPrioMsgPart.bind(_me);
+	  _me._running = true;
+	  _me._sendHitsBind();
+	  Event.observe(window, 'beforeunload', _me._sendMissedHits.bind(_me));
```
Comment (ebeutler) [inline src/main/webapp/resources/celRes/analyticsQueue.js]:
```suggestion
      _me._analyticsEventQueue = new Array();
      _me._analyticsPriorizedEventQueue = new Array();
      _me._addBind = _me.add.bind(_me);
      _me._priorityAddBind = _me.priorityAdd.bind(_me);
      _me._sizeBind = _me.size.bind(_me);
      _me._lowPrioritySizeBind = _me.lowPrioritySize.bind(_me);
      _me._highPrioritySizeBind = _me.highPrioritySize.bind(_me);
      _me._sendHitsBind = _me._sendHits.bind(_me);
      _me._getPrioMsgPartBind = _me._getPrioMsgPart.bind(_me);
      _me._running = true;
      _me._sendHitsBind();
      Event.observe(window, 'beforeunload', _me._sendMissedHits.bind(_me));
```

**Review by ebeutler — CHANGES_REQUESTED** [review]

File: src/main/webapp/resources/celDynJS/analytics/analyticsQueue.mjs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        const nextHit = this.#analyticsPriorizedEventQueue.shift();
+        window._paq.push(['trackEvent', nextHit.params.eventCategory,  nextHit.params.eventAction,
+         nextHit.params.eventLabel, nextHit.params.eventValue]);
+      }
+      while (this.lowPrioritySize() > 0) {
+        const nextHit = this.#analyticsEventQueue.shift();
+        window._paq.push(['trackEvent', nextHit.params.eventCategory,  nextHit.params.eventAction,
+         nextHit.params.eventLabel, nextHit.params.eventValue]);
+      }
+  }
```
Comment (msladek) [inline src/main/webapp/resources/celDynJS/analytics/analyticsQueue.mjs]:
```suggestion
  _sendHitsOnce() {
    while (this.highPrioritySize() > 0) {
      const nextHit = this.#analyticsPriorizedEventQueue.shift();
      window._paq.push(['trackEvent', nextHit.params.eventCategory, nextHit.params.eventAction,
        nextHit.params.eventLabel, nextHit.params.eventValue]);
    }
    while (this.lowPrioritySize() > 0) {
      const nextHit = this.#analyticsEventQueue.shift();
      window._paq.push(['trackEvent', nextHit.params.eventCategory, nextHit.params.eventAction,
        nextHit.params.eventLabel, nextHit.params.eventValue]);
    }
  }
```

File: src/main/webapp/resources/celDynJS/analytics/analyticsQueue.mjs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  _sendMissedHits() {
+    this.#running = false;
+    this._sendHitsOnce();
+    if ((this.highPrioritySize() + this.size()) > 0) {
```
Comment (msladek) [inline src/main/webapp/resources/celDynJS/analytics/analyticsQueue.mjs]:
```suggestion
    if (this.size() > 0) {
```
`size()` already contains `highPrioritySize()`

File: src/main/webapp/resources/celDynJS/analytics/analyticsQueue.mjs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+  async _sendHits() {
+    console.debug('send hits, #running', this.#running, ', queue size is', this.size());
+    while (this.#running) {
+      this._sendHitsOnce();
+      await this.#sleep(.5);
+    }
+  }
+
+  _sendHitsOnce() {
```
Comment (msladek) [inline src/main/webapp/resources/celDynJS/analytics/analyticsQueue.mjs]: change the prefix for private functions to `#`

**Review by msladek — CHANGES_REQUESTED** [review]

**Review by ebeutler — APPROVED** [review]

**Review by msladek — APPROVED** [review]

**Pull request merged**